File: care/emr/models/medication_administration.py
from datetime import datetime

from django.db import models

from care.emr.models.base import EMRBaseModel


class MedicationAdministration(EMRBaseModel):
    status = models.CharField(max_length=100)
    status_reason = models.JSONField(null=True, blank=True)
    category = models.CharField(max_length=100, null=True, blank=True)
    medication = models.JSONField(default=dict)
    patient = models.ForeignKey("emr.Patient", on_delete=models.CASCADE)
    encounter = models.ForeignKey(
        "emr.Encounter", on_delete=models.CASCADE, null=True, blank=True
    )
    request = models.ForeignKey(
        "emr.MedicationRequest", on_delete=models.CASCADE, null=True, blank=True
    )
    authored_on = models.DateTimeField(null=True, blank=True)
    occurrence_period_start = models.DateTimeField(default=datetime.now)
    occurrence_period_end = models.DateTimeField(null=True, blank=True)
    recorded = models.DateTimeField(null=True, blank=True)
    performer = models.JSONField(default=list)
    dosage = models.JSONField(null=True, blank=True)
    note = models.TextField(null=True, blank=True)
    administered_product = models.ForeignKey(
        "emr.ProductKnowledge", on_delete=models.CASCADE, null=True, blank=True
    )

File: frontend/src/types/emr/medicationAdministration/medicationAdministration.ts
import { Code } from "@/types/base/code/code";
import { DosageQuantity } from "@/types/emr/medicationRequest/medicationRequest";
import { ProductKnowledgeBase } from "@/types/inventory/productKnowledge/productKnowledge";
import { Quantity } from "@/types/questionnaire/quantity";
import { UserReadMinimal } from "@/types/user/user";

export const MEDICATION_ADMINISTRATION_STATUS = [
  "completed",
  "not_done",
  "entered_in_error",
  "stopped",
  "in_progress",
  "on_hold",
  "cancelled",
] as const;

export type MedicationAdministrationStatus =
  (typeof MEDICATION_ADMINISTRATION_STATUS)[number];

export interface MedicationAdministration {
  readonly id?: string;
  status: MedicationAdministrationStatus;
  status_reason?: Code;
  category?: "inpatient" | "outpatient" | "community";

  medication?: Code;
  administered_product?: string;

  authored_on?: string; // datetime
  occurrence_period_start: string; // datetime
  occurrence_period_end?: string; // datetime
  recorded?: string; // datetime

  encounter: string; // uuid
  request: string; // uuid

  performer?: {
    actor: string; // uuid
    function: "performer" | "verifier" | "witness";
  }[];
  dosage?: {
    text?: string;
    site?: Code;
    route?: Code;
    method?: Code;
    dose?: DosageQuantity;
    rate?: Quantity;
  };

  note?: string;

  created_by?: UserReadMinimal;
  updated_by?: UserReadMinimal;
}

export interface MedicationAdministrationRequest {
  id?: string;
  encounter: string;
  request: string;
  status: MedicationAdministrationStatus;
  status_reason?: Code;
  medication?: Code;
  administered_product?: string;
  occurrence_period_start: string;
  occurrence_period_end?: string;
  recorded?: string;
  note?: string;
  dosage?: {
    text?: string;
    site?: Code;
    route?: Code;
    method?: Code;
    dose?: DosageQuantity;
    rate?: Quantity;
  };
}

export interface MedicationAdministrationRead {
  id: string;
  status: MedicationAdministrationStatus;
  status_reason?: Code;
  medication?: Code;
  administered_product?: ProductKnowledgeBase;
  occurrence_period_start: string;
  occurrence_period_end?: string;
  recorded?: string;
  encounter: string;
  request: string;
  note?: string;
  dosage?: {
    text?: string;
    site?: Code;
    route?: Code;
    method?: Code;
    dose?: DosageQuantity;
    rate?: Quantity;
  };
  created_by: UserReadMinimal;
  updated_by: UserReadMinimal;
}
